unify reduction types from different operators: scatter, scatter_reduce, segment_reduce

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91501
* #91500
* __->__ #91499
* #90978

The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
unify reduction types from different operators: scatter, scatter_reduce, segment_reduce

[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"






[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"






cc jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"


The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"


The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"


The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"


The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

[ghstack-poisoned]

ngimel reviewed (approved):
This is search and replace mostly?

mingfeima commented:
> This is search and replace mostly?

Yes, this one is search and replace. There are other PRs with optimizations which will depend on this, such as `scatter_reduce`, `segment_reduce` etc. So that I can reuse the vectorization utils.

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"


The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

[ghstack-poisoned]

mingfeima committed:
Update on "unify reduction types from different operators: scatter, scatter_reduce, segment_reduce"


The target of this PR is to unify `ReductionType` for reduce operators so that we have the same set of reduce utils for `init`, or `update` for vectorization.

[ghstack-poisoned]

mingfeima commented:
@pytorchbot merge

malfet commented:
Hmm, marking this one as `BC-breaking` as it renames reduction types from "max" to "amax"

Also, not sure, why this was needed?

ngimel commented:
@mingfeima changing internal enums to be more uniform is fine, however, there's not enough reason to break user-facing APIs (and even if we did decide to break them, we should follow the process). Please revert changes to user-facing APIs.

malfet commented:
I will submit a forward fix, but I want to better understand the reason behind the rename

malfet commented on torch/masked/_ops.py:
```diff
@@ -783,9 +783,6 @@ def _sparse_csr_segment_reduction_helper(
             )
             new_nnz = new_crow_indices[-1]
             new_col_indices = col_indices.new_zeros(new_nnz)
-            # segment_reduce takes 'max'/'min' rather than 'amax'/'amin', changing this would be BC-breaking
```
I.e. this comment clearly says: don't change that!!!!!
